### PR TITLE
Improve argument printing

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #87294
* #87604
* #87603
* #87608
* __->__ #87601

No more "expected tuple but got tuple".  We appropriately
grovel in the list/tuple for the element that mismatched
and report what exactly twinged the failure.

invalid_arguments.cpp is a shitshow so I did something
slapdash to get it not completely horrible.  See
https://github.com/pytorch/pytorch/issues/87514 for more context.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Comment by albanD** [comment]
@ezyang added a couple of reinterpret_steal. Can you double check it is ok now?

File: torch/csrc/utils/invalid_arguments.cpp [hunk 1]
```diff
@@ -272,7 +272,32 @@ std::string _formattedArgDesc(
...
+      if (is_tuple) {
+        result += "(";
+      } else {
+        result += "[";
+      }
+      for (const auto i : c10::irange(num_elements)) {
+        if (i != 0) {
+          result += ", ";
+        }
+        result += py_typename(py::reinterpret_steal<py::object>(PySequence_GetItem(arg, i)));
```
Comment (ezyang) [inline torch/csrc/utils/invalid_arguments.cpp:288]: OOPS good catch

**Comment by ezyang** [comment]
Yes, very good, I accidentally thought GetItem returned a borrow.

**Review by Chillee — APPROVED** [review]
Makes sense to me and seems like a good change :)

I didn't read the logic of how we're interacting with Python carefully.

File: torch/csrc/utils/invalid_arguments.cpp (hunk 1 above)
Comment (albanD) [reply torch/csrc/utils/invalid_arguments.cpp]: Yes PySequence is the only one that returns a fresh reference.

File: torch/csrc/utils/python_arg_parser.cpp [hunk 2]
```diff
@@ -1374,6 +1397,24 @@ bool FunctionSignature::parse(
...
+              "%s(): argument '%s' (position %ld) must be %s, but found element of type %s at pos %ld",
+              name.c_str(),
+              param.name.c_str(),
+              static_cast<long>(arg_pos + 1),
+              param.type_name().c_str(),
+              Py_TYPE(py::reinterpret_steal<py::object>(
+                          PySequence_GetItem(obj, failed_idx))
+                          .ptr())
+                  ->tp_name,
+              static_cast<long>(failed_idx));
```
Comment (albanD) [inline torch/csrc/utils/python_arg_parser.cpp:1416]: Added a static cast here to fix arm64 compilation

**Pull request closed without merging**